src/System.Private.CoreLib/src/System/Diagnostics/Eventing/EventPipeController.cs

```diff
@@ -0,0 +1,291 @@
+        private const int DisabledPollingIntervalMilliseconds = 10000; // 10 seconds
+        private const uint DefaultCircularBufferMB = 1024; // 1 GB
+        private static readonly char[] ProviderConfigDelimiter = new char[] { ',' };
+        private static readonly char[] ConfigComponentDelimiter = new char[] { ':' };
+
+        // The default set of providers/keywords/levels.  Used if an alternative configuration is not specified.
+        private static readonly EventPipeProviderConfiguration[] DefaultProviderConfiguration = new EventPipeProviderConfiguration[]
+        {
+            new EventPipeProviderConfiguration("Microsoft-Windows-DotNETRuntime", 0x4c14fccbd, 5),
+            new EventPipeProviderConfiguration("Microsoft-Windows-DotNETRuntimePrivate", 0x4002000b, 5),
+            new EventPipeProviderConfiguration("Microsoft-DotNETCore-SampleProfiler", 0x0, 5)
+        };
+
+        // Cache for COMPlus configuration variables.
+        private static int s_Config_EnableEventPipe = -1;
+        private static string s_Config_EventPipeConfig = null;
+        private static uint s_Config_EventPipeCircularMB = 0;
+        private static string s_Config_EventPipeOutputFile = null;
+
+        // Singleton controller instance.
+        private static EventPipeController s_controllerInstance = null;
+
+        // Controller object state.
+        private string m_traceFilePath;
+        private string m_markerFilePath;
+        private bool m_markerFileExists;
+
+        internal static void Initialize()
+        {
+            // Don't allow failures to propagate upstream.
+            // Instead, ensure program correctness without tracing.
+            try
+            {
+                if (s_controllerInstance == null)
+                {
+                    if(Config_EnableEventPipe == 4)
+                    {
+                        // Create a new controller to listen for commands.
+                        s_controllerInstance = new EventPipeController();
+                    }
+                    else if (Config_EnableEventPipe > 0)
+                    {
+                        // Enable tracing immediately.
+                        // It will be disabled automatically on shutdown.
+                        EventPipe.Enable(GetConfiguration());
+                    }
+                }
+            }
+            catch { }
+        }
+
+        private EventPipeController()
+        {
+            // Set file paths.
+            m_traceFilePath = GetDisambiguatedTraceFilePath(Config_EventPipeOutputFile);
+            m_markerFilePath = MarkerFilePath;
+
+            // Marker file is assumed to not exist.
+            // This will be updated when the monitoring thread starts.
+            m_markerFileExists = false;
+
+            // Start a new thread to listen for tracing commands.
+            Task.Factory.StartNew(WaitForTracingCommand, TaskCreationOptions.LongRunning);
+        }
```

Comment:
I am including this again, because it may have been lost inadvertantly.  

We can avoid burning the thread by
```
 m_timer = new Timer(PollTracingCommand, null, 10000, 10000);
```
WHere PollForTracingCommand is now one iteration of the loop in WaitForTracingCommand.  

This achieves the same effect but does not tie up a thread for the duration of the process.

Reply:
Yes, this was lost inadvertently.  I have switched to a timer.  Please take a look at the implementation.  I've set it up to avoid concurrent polling operations by scheduling the next poll operation during the callback.  I've done some measurement and have not seen this show up, but of course, this is with a sampling profiler.  Let me know if you have any concerns with this.

Reply:
I have no concern with scheduling the next timer in the callback.   A 10 second interval insures that the cost will be << .1%.